adapters/insticator/insticator.go

```diff
@@ -0,0 +1,335 @@
+	var errs []error
+	var adapterRequests []*adapters.RequestData
+	var groupedImps = make(map[string][]openrtb2.Imp)
+
+	reqExt, err := makeReqExt(request)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	request.Ext = reqExt
+
+	// Create a deep copy of the request to avoid modifying the original request
+	requestCopy := *request
+	requestCopy.Site = request.Site
+	requestCopy.App = request.App
+
+	for i := 0; i < len(request.Imp); i++ {
+		impCopy, impKey, err := makeImps(request.Imp[i])
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		// Populate site.publisher.id from imp extension
+		if requestCopy.Site != nil || requestCopy.App != nil {
+			if err := populatePublisherId(&impCopy, &requestCopy); err != nil {
+				errs = append(errs, err)
+				continue
+			}
+		}
+
+		resolvedBidFloor, errFloor := resolveBidFloor(impCopy.BidFloor, impCopy.BidFloorCur, requestInfo)
+		if errFloor != nil {
+			errs = append(errs, &errortypes.BadInput{
+				Message: fmt.Sprintf("Error in converting the provided bid floor currency from %s to USD",
+					impCopy.BidFloorCur),
+			})
+			continue
+		} else {
+			if resolvedBidFloor > 0 {
+				impCopy.BidFloor = resolvedBidFloor
+				impCopy.BidFloorCur = "USD"
+			}
+		}
+
+		groupedImps[impKey] = append(groupedImps[impKey], impCopy)
+	}
+
+	for _, impList := range groupedImps {
+		if adapterReq, err := a.makeRequest(requestCopy, impList); err == nil {
+			adapterRequests = append(adapterRequests, adapterReq)
+		} else {
+			errs = append(errs, err)
+		}
+	}
+	return adapterRequests, errs
+}
+
+func (a *adapter) makeRequest(request openrtb2.BidRequest, impList []openrtb2.Imp) (*adapters.RequestData, error) {
+	request.Imp = impList
+
+	reqJSON, err := json.Marshal(request)
+	if err != nil {
+		return nil, err
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+
+	if request.Device != nil {
+		if len(request.Device.UA) > 0 {
+			headers.Add("User-Agent", request.Device.UA)
+		}
+
+		if len(request.Device.IPv6) > 0 {
+			headers.Add("X-Forwarded-For", request.Device.IPv6)
+		}
+
+		if len(request.Device.IP) > 0 {
+			headers.Add("X-Forwarded-For", request.Device.IP)
+			headers.Add("IP", request.Device.IP)
+		}
+	}
+
+	return &adapters.RequestData{
+		Method:  "POST",
+		Uri:     a.endpoint,
+		Body:    reqJSON,
+		Headers: headers,
+		ImpIDs:  openrtb_ext.GetImpIDs(request.Imp),
+	}, nil
+}
+
+func (a *adapter) MakeBids(request *openrtb2.BidRequest, requestData *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if adapters.IsResponseStatusCodeNoContent(responseData) {
+		return nil, nil
+	}
+
+	if err := adapters.CheckResponseStatusCodeForErrors(responseData); err != nil {
+		return nil, []error{err}
+	}
+
+	var response openrtb2.BidResponse
+	if err := json.Unmarshal(responseData.Body, &response); err != nil {
```

Comment:
Please use our jsonutil helper as a drop in replacement for json. This is a new policy as of PBS 3.0 for performance.